data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetStateValidators.java

```diff
@@ -0,0 +1,134 @@
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.teku.api.ChainDataProvider;
+import tech.pegasys.teku.api.DataProvider;
+import tech.pegasys.teku.api.response.v1.beacon.GetStateValidatorsResponse;
+import tech.pegasys.teku.beaconrestapi.ListQueryParameterUtils;
+import tech.pegasys.teku.beaconrestapi.handlers.AbstractHandler;
+import tech.pegasys.teku.beaconrestapi.schema.BadRequest;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.provider.JsonProvider;
+import tech.pegasys.teku.storage.client.ChainDataUnavailableException;
+
+public class GetStateValidators extends AbstractHandler {
+  private static final Logger LOG = LogManager.getLogger();
+  public static final String ROUTE = "/eth/v1/beacon/states/:state_id/validators";
+
+  private final ChainDataProvider provider;
+
+  public GetStateValidators(final DataProvider dataProvider, final JsonProvider jsonProvider) {
+    this(dataProvider.getChainDataProvider(), jsonProvider);
+  }
+
+  GetStateValidators(final ChainDataProvider provider, final JsonProvider jsonProvider) {
+    super(jsonProvider);
+    this.provider = provider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get validators from state",
+      tags = {TAG_V1_BEACON, TAG_VALIDATOR_REQUIRED},
+      description = "Returns filterable list of validators with their balance, status and index.",
+      pathParams = {
+        @OpenApiParam(name = PARAM_STATE_ID, description = PARAM_STATE_ID_DESCRIPTION),
+      },
+      queryParams = {
+        @OpenApiParam(
+            name = PARAM_VALIDATOR_ID,
+            description = PARAM_VALIDATOR_DESCRIPTION,
+            isRepeatable = true)
+      },
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = @OpenApiContent(from = GetStateValidatorsResponse.class)),
+        @OpenApiResponse(status = RES_BAD_REQUEST),
+        @OpenApiResponse(status = RES_INTERNAL_ERROR),
+        @OpenApiResponse(status = RES_SERVICE_UNAVAILABLE, description = SERVICE_UNAVAILABLE)
+      })
+  @Override
+  public void handle(@NotNull final Context ctx) throws Exception {
+    try {
+      final UInt64 slot = parseSlotParam(ctx);
+
+      final List<Integer> validatorIndices = parseValidatorsParam(ctx);
+
+      SafeFuture<Optional<GetStateValidatorsResponse>> future =
+          provider
+              .getValidatorsDetails(slot, validatorIndices)
+              .thenApply(result -> result.map(GetStateValidatorsResponse::new));
+
+      ctx.header(Header.CACHE_CONTROL, getMaxAgeForSlot(provider, slot));
+      if (provider.isFinalized(slot)) {
+        handlePossiblyGoneResult(ctx, future);
+      } else {
+        handlePossiblyMissingResult(ctx, future);
+      }
+    } catch (ChainDataUnavailableException ex) {
+      LOG.trace(ex);
+      ctx.status(SC_SERVICE_UNAVAILABLE);
```

Comment:
you'll need to set the result here too, like below, `BadMessage.serviceUnavailable` was just added, should make that trivial

Reply:
Done. Also updated the single validator version that I based this off.